exercises/game-of-life/canonical-data.json

```diff
@@ -0,0 +1,75 @@
+{
+  "exercise": "game-of-life",
+  "cases": [
+    {
+      "uuid": "ae86ea7d-bd07-4357-90b3-ac7d256bd5c5",
+      "description": "empty matrix",
+      "property": "gameOfLife",
+      "input": {
+        "matrix": []
+      },
+      "expected": []
+    },
+    {
+      "uuid": "4ea5ccb7-7b73-4281-954a-bed1b0f139a5",
+      "description": "dead cells with three live neighbors become alive",
+      "property": "gameOfLife",
+      "input": {
+        "matrix": [
+          [1, 1, 0],
+          [0, 0, 0],
+          [1, 0, 0]
+        ]
+      },
+      "expected": [
+        [0, 0, 0],
+        [1, 1, 0],
+        [0, 0, 0]
+      ]
+    },
+    {
+      "uuid": "7cf0afcd-995c-4bc8-b71e-9deeef532f64",
+      "description": "overpopulated cells",
+      "property": "gameOfLife",
+      "input": {
+        "matrix": [
+          [1, 1, 1],
+          [1, 0, 1],
+          [1, 1, 1]
+        ]
+      },
+      "expected": [
+        [1, 0, 1],
+        [0, 0, 0],
+        [1, 0, 1]
+      ]
+    },
+    {
+      "uuid": "a79b42be-ed6c-4e27-9206-43da08697ef6",
+      "description": "random matrix",
+      "property": "gameOfLife",
+      "input": {
+        "matrix": [
+          [1, 1, 0, 1, 1, 0, 0, 0],
+          [1, 0, 1, 1, 0, 0, 0, 0],
+          [1, 1, 1, 0, 0, 1, 1, 1],
+          [0, 0, 0, 0, 0, 1, 1, 0],
+          [1, 0, 0, 0, 1, 1, 0, 0],
+          [1, 1, 0, 0, 0, 1, 1, 1],
+          [0, 0, 1, 0, 1, 0, 0, 1],
+          [1, 0, 0, 0, 0, 0, 1, 1]
+        ]
+      },
+      "expected": [
+        [1, 1, 0, 1, 1, 0, 0, 0],
+        [0, 0, 0, 0, 0, 1, 1, 0],
+        [1, 0, 1, 1, 1, 1, 0, 1],
+        [1, 0, 0, 0, 0, 0, 0, 1],
+        [1, 1, 0, 0, 1, 0, 0, 1],
+        [1, 1, 0, 1, 0, 0, 0, 1],
+        [1, 0, 0, 0, 0, 0, 0, 0],
+        [0, 0, 0, 0, 0, 0, 1, 1]
+      ]
```

Comment:
In order to fully specify the basic rules of the cellular automata, I think we should have at least the following tests:

1. Underpopulation: live cell that dies due to having no live neighbors
2. Underpopulation: live cell that dies due to having only one live neighbor.
3. Survival: A live cell that survives to the next generation due to having two live neighbors.
4. Survival: A live cell that survives to the next generation due to having three live neighbors.
5. Overcrowding: A live cell that dies due to having four live neighbors.
6. Reproduction: A dead cell with exactly three live neighbors becomes a live cell.

In each of these, there would only be one turn.

I would instinctively want to have individual tests for a live cell that dies due to overcrowding with five, six, seven, and eight live neighbors, but I don't know this problem well, so I can't tell if this would be overkill.

Ideally each of these tests would have the smallest possible matrix with the fewest possible changes.

Reply:
I think this is a very good idea. I think it would be great to see a test for each count of neighbors. Maybe with an empty border so there's no literal edge cases. 

Would it also make sense to have a grid which combines all those individual minimal unit tests into a single test?

Reply:
> Would it also make sense to have a grid which combines all those individual minimal unit tests into a single test?

That would be a puzzle:

| Alive neighbors | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 |
|--------|-|-|-|-|-|-|-|-|-|
| Cases | 1 | 8 | 28 | 56 | 70 | 56 | 28 | 8 | 1 |

(Multiply by 2 to account for the focus being alive/dead.)

Also, it is impossible to have test cases that _only_ check certain numbers of alive neighbors, for all numbers. (But is is very much possible for _some_.)

Reply:
It doesn't have to be perfect. 

```
# Zero
000
000
000

# One
100
000
000

# Two
110
000
000

# Three 
101
000
001

# Etc
```

Reply:
For combining, I meant literally having a larger matrix which has the minimal matrices copy pasted into non overlapping sections.

Reply:
IMO we don't need an Overcrowding test for every number, but the test I added for it naturally has different numbers of neighbors for each cell.

Reply:
Could we rename this test? "Random" data doesn't make for a good test. Maybe "larger board" or "bigger test"? If it's random, what is it actually testing?